### PR TITLE
Adds support for functions that return non-introspected subclasses of…

### LINKED ISSUE #74: Support functions that return non-introspected subclasses of introspected types
```js
const gi = require('node-gtk')
const Gtk = gi.require('Gtk', '3.0')
const Gdk = gi.require('Gdk', '3.0')

gi.startLoop()
Gtk.init()
Gdk.init([])


const css = new Gtk.CssProvider()
// css.loadFromPath(path.join(__dirname, 'style.css'))
css.loadFromData(`
button {
  padding: 20px;
}
`.split('').map(s => s.charCodeAt(0)))
const screen = Gdk.Display.getDefaultScreen(Gdk.Display.getDefault())
Gtk.StyleContext.addProviderForScreen(screen, css, 600)
```
Fails with: `gobject.cc: GetClassTemplate: 380: Assertion 'gi_info != NULL' failed: Missing GIR info for: GdkX11Display`

### PR DESCRIPTION
**Pull request by k0d**
Fixes #74 
I've added an example gst.js to test my changes. It 'should' show a test video.

Note: I've force pushed a proper commit that is tested and works. It does indeed fix issue #74 but in a different way to the way I've seen you do it in your new branch. What do you think is best to do?

### CODE REVIEW

File: examples/gst.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const GiJS = require('../lib/')
```
Comment (romgrk) [inline examples/gst.js]: Keep it consistent with the rest, use `gi` as the variable name. If you have a better name, still follow JS conventions (lowerCamelCase for variables).

File: lib/index.js [hunk 2]
```diff
@@ -568,6 +568,8 @@ function prependLibraryPath(path) {
 // Public API
 exports.require = giRequire
 exports.startLoop = internal.StartLoop
+exports.setProperty = internal.ObjectPropertySetter
+exports.getProperty = internal.ObjectPropertyGetter
```
Comment (romgrk) [inline lib/index.js]: Public API requires documentation. For the names, I would prefer `(s|g)etObjectProperty`.

File: src/gobject.cc [hunk 3]
```diff
@@ -400,7 +400,7 @@ Local<Function> MakeClass(GIBaseInfo *info) {
     return tpl->GetFunction ();
 }
 
-Local<Value> WrapperFromGObject(GObject *gobject) {
+Local<Value> WrapperFromGObject(GObject *gobject, GIBaseInfo *object_info) {
```
Comment (romgrk) [inline src/gobject.cc]: Why is this changed if it's not used anywhere?

File: examples/gst.js (hunk 1 above)
Comment (k0d) [reply examples/gst.js]: Fixed

File: lib/index.js (hunk 2 above)
Comment (k0d) [reply lib/index.js]: Will add this into a different issue/PR

File: src/gobject.cc (hunk 3 above)
Comment (k0d) [reply src/gobject.cc]: It is now :) That was the missing code.

**Comment by k0d** [comment]
I know there are 2 issues, and the code in this commit solves #74 
The properties issue, that's a different issue #83 which I've not committed code to yet. There was a hack fix which was in the commit earlier...but isn't now.

File: src/gobject.h [hunk 4]
```diff
@@ -15,7 +15,7 @@ using v8::Value;
 namespace GNodeJS {
 
 Local<Function>         MakeClass            (GIBaseInfo *info);
-Local<Value>            WrapperFromGObject   (GObject *object);
+Local<Value>            WrapperFromGObject   (GObject *object, GIBaseInfo *info);
```
Comment (romgrk) [inline src/gobject.h]:
```suggestion
Local<Value>            WrapperFromGObject   (GObject *object, GIBaseInfo *info = NULL);
```

File: src/value.cc [hunk 5]
```diff
@@ -1195,7 +1195,7 @@ Local<Value> GValueToV8(const GValue *gvalue) {
     } else if (G_VALUE_HOLDS_ENUM (gvalue)) {
         return New<Integer>(g_value_get_enum (gvalue));
     } else if (G_VALUE_HOLDS_OBJECT (gvalue)) {
-        return WrapperFromGObject (G_OBJECT (g_value_get_object (gvalue)));
+        return WrapperFromGObject (G_OBJECT (g_value_get_object (gvalue)), NULL);
```
Comment (romgrk) [inline src/value.cc]:
```suggestion
        return WrapperFromGObject (G_OBJECT (g_value_get_object (gvalue)));
```

File: src/gobject.cc [hunk 6]
```diff
@@ -416,7 +416,8 @@ Local<Value> WrapperFromGObject(GObject *gobject) {
         GType gtype = G_OBJECT_TYPE(gobject);
         g_type_ensure (gtype); //void *klass = g_type_class_ref (type);
 
-        auto tpl = GetClassTemplate(NULL, gtype);
+        // We don't use the gtype above, but maybe we can register that type using the type's interface's object_info.
```
Comment (romgrk) [inline src/gobject.cc]: Then we can remove unused variables (`gtype`).

**Comment by romgrk** [comment]
Ping @k0d 

I haven't had time to inspect this further, so if you'd like please review & fix the comments above and I think we'll be good to merge it.

**Comment by k0d** [comment]
I’ve been sick the past days so didn’t look at this either recently. However I have fixed some other issues and committed them to my work’s fork. They will likely break the testing and might not ve compatible with your environment. I plan to do the changes in a more portable way and make a pull request in a few days. I understand the whole Nan and GIR systems much better now, have learnt a lot recently!

**Comment by romgrk** [comment]
Awesome :) Send along the new changes whenever you're ready.
I've also learned a lot working on this project!, there are a lot of moving parts to grasp.

**Comment by romgrk** [comment]
Ping @kod! It happens that I ran into this issue with one project I'm working on, so I'd be happy to have this merged as soon as possible. Would you like to address the 3 suggestions above, and then I'll merge? Thanks!

**Comment by k0d** [comment]
I've actually written gobject introspection bindings for nodeJS from scratch using NAPI, implementing any missing features and fixing any known issues, I've not decided yet what I'll do with it...I'm not working at that job anymore and have written this 100% in my own time so I can license it however I want. I'm thinking MIT...

I don't have time to work on this project (node-gtk) right now, would you be interested in the other project? If I open it up, then I'd really appreciate some bug-finding.

**Comment by romgrk** [comment]
Well I'm happy with node-gtk, it does what I need it to do for now. 
I would be interested in that project, but only if it's in a good enough state to make this one irrelevant. 

My other problem is that for my current usage, I need `GtkDrawingArea` capabilities, but the `cairo` library is not introspectable, which means it needs to be manually binded (#93).

So basically no, but I'd be happy to see it published, just in case :)
